bundle/jsky.app.ot/src/main/scala/jsky/app/ot/ags/BagsManager.scala

```diff
@@ -25,219 +24,396 @@ import jsky.app.ot.gemini.tpe.TpePWFSFeature
 import jsky.app.ot.tpe.{TpeContext, TpeManager}
 
 import scala.collection.JavaConverters._
-import scala.concurrent.{ExecutionContext, Future}
+import scala.concurrent.ExecutionContext
 import scala.swing.Swing
 import scala.util.{Failure, Success}
 import scalaz._
 import Scalaz._
+import scalaz.effect.IO
+import scalaz.effect.IO.ioUnit
 
+/** The state of AGS lookup for a single observation. */
+sealed trait BagsState {
+  /** Called when an observation is edited to determine what should be the new
+    * state.  This is typically called in response to a listener on the
+    * science program.
+    */
+  def edit: StateTransition =
+    ErrorTransition
 
-final class BagsManager(executorService: ExecutorService) {
-  private val LOG: Logger = Logger.getLogger(getClass.getName)
+  /** Called when a timer goes off to evaluate what should be done next and
+    * possibly move to a new state.  Timer events transition out of pending
+    * into running or out of failed into pending.  When an AGS lookup fails,
+    * a few seconds are added before transitioning in order to provide an
+    * opportunity for network issues, etc. to sort themselves out.
+    */
+  def wakeUp: StateTransition =
+    (this, ioUnit)
 
-  private implicit val executionContext = new ExecutionContext {
-    def shutdown() = executorService.shutdown()
-    override def reportFailure(t: Throwable): Unit = throw t
-    override def execute(runnable: Runnable): Unit = {
-      executorService.submit(new Thread(runnable) {
-        setPriority(Thread.NORM_PRIORITY - 1)
-      })
-    }
-  }
+  /** Called with the successful results of an AGS lookup. */
+  def succeed(results: Option[AgsStrategy.Selection]): StateTransition =
+    ErrorTransition
 
-  /** Our state is just a set of programs to watch, and a set of pending keys. */
-  private case class BagsState(programs: Set[SPProgramID], keys: Set[SPNodeKey]) {
-    def +(key: SPNodeKey): BagsState = copy(keys = keys + key)
-    def -(key: SPNodeKey): BagsState = copy(keys = keys - key)
-    def +(pid: SPProgramID): BagsState = copy(programs = programs + pid)
-    def -(pid: SPProgramID): BagsState = copy(programs = programs - pid)
-  }
-  @volatile private var state: BagsState = BagsState(Set.empty, Set.empty)
+  /** Called to notify that an AGS lookup failed. */
+  def fail(why: String, delayMs: Long): StateTransition =
+    ErrorTransition
+}
 
-  // Setup an LRU cache with max entries BagsManager.CacheSize
-  private val hashes = new util.LinkedHashMap[SPNodeKey, Int](math.ceil((BagsManager.CacheSize + 1) * 1.334).toInt, 0.75f, true) {
-    override def removeEldestEntry(e: java.util.Map.Entry[SPNodeKey, Int]): Boolean =
-      size > BagsManager.CacheSize
+
+object BagsState {
+  type AgsHashVal      = Long
+  type StateTransition = (BagsState, IO[Unit]) // Next BagsState, side-effects
+  val ErrorTransition  = (ErrorState, ioUnit)  // Default transition for unexpected events
```

Comment:
It's "an `IO` action that does nothing".  A constant value for `IO(())`.  Here we need an `IO` action to conform to the `StateTransition` type but we don't actually have any actions to take.